File: leetcode/lowest-common-ancestor-of-a-binary-tree.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <stack>
#include <queue>
#include "../utils.h"
#include "../tree.h"

using namespace std;

// https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

//      3(l, r)
//   5      1
// 6   2   0  8
//   7   4
// p=5, q=1

// DFS (postorder)
// 3(t) 1(t)
// time: o(n)
// space: o(n) recursion

//      3(l, r)
//   5      1
// 6   2   0  8
//   7   4
// p=5, q=4

// DFS
// 3, 5(t) 2, 4(t)

// TreeNode* ans;
// bool lca(root, p, q)
//      if (root == nullptr) return 0;
//      int left = lca(root->left, p, q);
//      int right = lca(root->right, p, q);
//      int mid = (root == p || root == q) ? 1 : 0;
//      if (mid + left + right >= 2)
//          ans = root;
//      return (mid + left + right > 0);
// lca(root, p, q)
// return ans;

class Solution {
private:
    TreeNode* ans;
    bool lca(TreeNode* node, TreeNode* p, TreeNode* q) {
        if (node == nullptr) return false;
        
        auto l = lca(node->left, p, q) ? 1 : 0;
        auto r = lca(node->right, p, q) ? 1 : 0;
        auto mid = (node == p || node == q) ? 1 : 0;
        if (l + r + mid >= 2) {
            this->ans = node;
        }
        
        return l + r + mid > 0;
    }
    
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        lca(root, p, q);
        return ans;
    }
};

int main()
{
    return 0;
}

File: leetcode/leetcode_py/lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

'''
 https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree

 Solution1. Build ChildParent map (child: parent, level) and find common parent
 time: o(n) n + n + n
 space: o(n);

 Solution2-3. DFS return pointer if it hits p or q (or null)
 time: o(n)
 space: o(n) height
 '''

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if not root or root == p or root == q: return root
        l = self.lowestCommonAncestor(root.left, p, q)
        r = self.lowestCommonAncestor(root.right, p, q)
        if l and r: return root
        return l if l else r

        